platform/xen/rumpxendev/evtdev.c

```diff
@@ -0,0 +1,502 @@
+		mutex_exit(&devevent_lock);
+		minios_bind_evtchn(bind_intd->port, minios_evtdev_handler, d);
+		minios_unmask_evtchn(bind_intd->port);
+
+		break;
+	}
+	case IOCTL_EVTCHN_BIND_UNBOUND_PORT:
+	{
+		struct ioctl_evtchn_bind_unbound_port *bind_unbound = data;
+		evtchn_alloc_unbound_t op;
+
+		op.dom = DOMID_SELF;
+		op.remote_dom = bind_unbound->remote_domain;
+		if ((err = minios_event_channel_op(EVTCHNOP_alloc_unbound, &op))) {
+			printf("IOCTL_EVTCHN_BIND_UNBOUND_PORT failed: "
+			       "remote domain %d error %d\n",
+			       bind_unbound->remote_domain, err);
+			return -err;
+		}
+		bind_unbound->port = op.port;
+		mutex_enter(&devevent_lock);
+		KASSERT(devevents[bind_unbound->port] == NULL);
+		devevents[bind_unbound->port] = d;
+		mutex_exit(&devevent_lock);
+		minios_bind_evtchn(bind_unbound->port, minios_evtdev_handler, d);
+		minios_unmask_evtchn(bind_unbound->port);
+
+		break;
+	}
+	case IOCTL_EVTCHN_UNBIND:
+	{
+		struct ioctl_evtchn_unbind *unbind = data;
+
+		if (unbind->port > NR_EVENT_CHANNELS)
+			return EINVAL;
+		mutex_enter(&devevent_lock);
+		if (devevents[unbind->port] != d) {
+			mutex_exit(&devevent_lock);
+			return ENOTCONN;
+		}
+		devevents[unbind->port] = NULL;
+		mutex_exit(&devevent_lock);
+		minios_mask_evtchn(unbind->port);
+		minios_unbind_evtchn(unbind->port);
+
+		break;
+	}
+	case IOCTL_EVTCHN_NOTIFY:
+	{
+		struct ioctl_evtchn_notify *notify = data;
+
+		if (notify->port > NR_EVENT_CHANNELS)
+			return EINVAL;
+		mutex_enter(&devevent_lock);
+		if (devevents[notify->port] != d) {
+			mutex_exit(&devevent_lock);
+			return ENOTCONN;
+		}
+		minios_notify_remote_via_evtchn(notify->port);
+		mutex_exit(&devevent_lock);
+
+		break;
+	}
+	default:
+		return EINVAL;
+	}
+
+	return 0;
+}
+
+/*----- setup etc. -----*/
+
+static int
+xenevt_dev_close(struct file *fp)
+{
+	struct xenevt_dev_data *d = fp->f_data;
+	int i;
+
+	DPRINTF(("/dev/xenevt: close...\n"));
+
+	mutex_enter(&devevent_lock);
+	mutex_enter(&d->lock);
+	for (i = 0; i < NR_EVENT_CHANNELS; i++) {
+		if (devevents[i] == d) {
+			minios_unbind_evtchn(i);
+			devevents[i] = NULL;
+		}
+	}
+	mutex_exit(&d->lock);
+	mutex_exit(&devevent_lock);
+
+	seldestroy(&d->selinfo);
+	mutex_destroy(&d->lock);
+	bmk_memfree(d);
+	cv_destroy(&d->cv);
+
+	DPRINTF(("/dev/xenevt: close done.\n"));
+
+	fp->f_data = NULL;
+
+	return 0;
+}
+
+const struct fileops xenevt_dev_fileops = {
+        .fo_read = xenevt_dev_read,
+        .fo_write = xenevt_dev_write,
+        .fo_ioctl = xenevt_dev_ioctl,
+        .fo_fcntl = fnullop_fcntl,
+        .fo_poll = xenevt_dev_poll,
+        .fo_stat = fbadop_stat,
+        .fo_close = xenevt_dev_close,
+        .fo_kqfilter = fnullop_kqfilter,
+        .fo_restart = fnullop_restart,
+};
+
+int
+xenevt_dev_open(struct file *fp, void **fdata_r)
+{
+	struct xenevt_dev_data *d;
+
+	d = bmk_memcalloc(1, sizeof(*d));
+	if (!d)
+		return ENOMEM;
+
+	memset(d, 0, sizeof(*d));
+
+	mutex_init(&d->lock, MUTEX_DEFAULT, IPL_HIGH);
```

Comment:
IPL_HIGH does nothing in a rump kernel.  Better just use `IPL_NONE` and not give false hope.